.github/workflows/release.yml

```diff
@@ -0,0 +1,92 @@
+name: Release
+on:
+  push:
+    tags:
+      - "v[0-9]+.[0-9]+.[0-9]+"
+      - "v[0-9]+.[0-9]+.[0-9]+rc[0-9]+"
+      - "v[0-9]+.[0-9]+.[0-9]+dev[0-9]+"
+  workflow_dispatch:
+    inputs:
+      skip-tests:
+        type: boolean
+        description: Skip
+        default: false
+      ref:
+        type: string
+        description: Tag to release
+        required: true
+
+permissions:
+  contents: read
+
+jobs:
+  pre-commit:
+    uses: ./.github/workflows/step_pre-commit.yml
+    if: "! inputs.skip-tests"
+
+  test-pip:
+    needs: [ pre-commit ]
+    uses: ./.github/workflows/step_tests-pip.yml
+    if: "! inputs.skip-tests"
+
+  test-conda:
+    needs: [ test-pip ]
+    uses: ./.github/workflows/step_tests-conda.yml
+    if: "! inputs.skip-tests"
+
+  test-ui:
+    needs: [ test-pip ]
+    uses: ./.github/workflows/step_tests-ui.yml
+    if: "! inputs.skip-tests"
+
+  test-status:
+    needs: [ pre-commit, test-pip, test-conda, test-ui ]
+    runs-on: ubuntu-latest
+    steps:
+      - uses: re-actors/alls-green@release/v1
+        with:
+          allowed-skips: pre-commit, test-pip, test-conda, test-ui
+          jobs: ${{ toJSON(needs) }}
+    if: always()
+
+  build:
+    needs: [ test-status ]
+    uses: ./.github/workflows/step_build.yml
+    with:
+      upload: true
+      ref: ${{ inputs.ref }}
+
+  publish:
+    needs: [ build ]
+    runs-on: ubuntu-latest
+
+    environment:
+      name: pypi
+      url: https://pypi.org/p/jupytext
+
+    permissions:
+      contents: read
+      id-token: write
+
+    steps:
+      - uses: actions/download-artifact@v3
+        with:
+          name: dist
+          path: dist
+      - name: Publish
+        uses: pypa/gh-action-pypi-publish@release/v1
+
+  release:
+    needs: [ publish ]
+    name: Create release
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          ref: ${{ inputs.ref }}
+      - uses: softprops/action-gh-release@v1
+        with:
+          name: Jupytext ${{ inputs.ref || github.ref_name }}
+          draft: true
+          prerelease: ${{ contains(inputs.ref || github.ref, 'rc') }}
+          generate_release_notes: true
```

Comment:
Instead of reinventing the wheel, I would suggest to look into jupyter-releaser. It will work out-of-the-box nicely to publish to both PyPI and NPM.

Reply:
Sure, open to suggestions on this one, can you propose the change so I can look at what it does and check if it's equivalent? I'm curious if it works with `Trusted publishing` and how that works when being used with both PyPI and NPM

Reply:
It does work with `Trusted publishing` which is very neat. For NPM we still need to use a token as a secret to publish.

Reply:
I need to read it carefully, because the workflow seems incompatible:
https://github.com/jupyter-server/jupyter_releaser/blob/43d510770f1fe039e133890936416956b927e250/jupyter_releaser/actions/populate_release.py#L42

(This workflow runs on tag release, while the `jupyter_releasser` creates the new tag) 🤔. Indeed there are good points like the bumpimg of `package.json`, but then how do you specify the tag?

Reply:
And we specify the tag in `Version specifier` input. As we use `hatch` already, different specifiers like `major, minor, patch, relase, build` are supported out-of-the box.

Reply:
Ok, I think we have two possible workflows here:
- run on tag: use the current approach and run `pre-release` to bump to the next version. I think we can mix individual steps from the releaser as long as we order them appropriately
- run on dispatch: use the reusable workflows there

It is a bit weird because of the mixing of dynamic-version via git on the python sideand the static version of npm. Normally I prefer to bump versions after a release in order to clearly communicate the difference in versions

Reply:
Well, my original suggestion was to replace the current `release.yml` file with the ones from [jupyter-releaser](https://github.com/jupyter-server/jupyter_releaser/tree/main/example-workflows) instead of mixing them with current one. But I see few potential blockers adopting `jupyter-releaser`:

- `jupyter-releaser` expects `package.json` to be in root of the repo for it to build and release NPM packages.
- And NPM package and Python package should have same version for consistency. For example, JupyterLab has few dozens of NPM packages and they all have consistent version with main JupyterLab Python package.

> dynamic-version via git on the python sideand the static version of npm

We do not use dynamic version (if you mean like `versioneer` creating a new version string for each commit). What hatch will do is bump NPM version in `package.json` and then make sure that Python package will use that same version as `package.json`. Currently, versions of NPM package and Python package of Jupytext are diverged and hence, it does not work out-of-the-box.

Your suggestion of creating our own workflows based on `jupyter-releaser` _might_ work. Something like splitting like [these workflows](https://github.com/jupyter-server/jupyter_releaser/tree/main/example-workflows) for Python and NPM packages separately. But that increases the maintainence burden. As @mwouts is the one who is making releases, it depends on how much complexity he is willing to handle.

Reply:
> Well, my original suggestion was to replace the current `release.yml` file with the ones from [jupyter-releaser](https://github.com/jupyter-server/jupyter_releaser/tree/main/example-workflows) instead

Slight nuance because you still need a `release.yml` with an `on:workflow_dispatch`. But yeah, what you suggest is about the second approach in my last comment. We just need to have a light wrapper to forward the inputs.

> We do not use dynamic version (if you mean like `versioneer` creating a new version string for each commit).

Ah you're right `src/jupytext/version.py` is a static file. I confused it with `tool.hatch.build.hooks.vcs.version-file`. Would the version bumping work on the python side?

> Currently, versions of NPM package and Python package of Jupytext are diverged and hence, it does not work out-of-the-box.

Good point. In either case it needs to re-converge the versioning number, unless there are more complicated tag forms.

> Your suggestion of creating our own workflows based on jupyter-releaser might work. Something like splitting like [these workflows](https://github.com/jupyter-server/jupyter_releaser/tree/main/example-workflows) for Python and NPM packages separately.

Indeed, looking back there are quite a few changes necessary. And in either case we need to test these workflows because there seems to be quite a few things to look out for. @mahendrapaipuri can the `jupyter-releaser` be tested locally on a fork, like uploading to GitHub package repo. And indeed the main question is @mwouts, which one do you prefer: tag push, calling workflow, or manual commits?

Reply:
> Ah you're right src/jupytext/version.py is a static file. I confused it with tool.hatch.build.hooks.vcs.version-file. Would the version bumping work on the python side?

Ah, I thought we are already using `hatch` to do version bumping. But currenctly version string is set manually I assume. Yes, we can do the bumping using `hatch` 

> can the jupyter-releaser be tested locally on a fork, like uploading to GitHub package repo

Yes, there is a [check-release](https://github.com/jupyter-server/jupyter_releaser/blob/main/.github/workflows/check-release.yml) workflow that dry runs all the release steps on local PyPI and mock GitHub instance in CI. So, if and when we decide to use `jupyter-releaser` tools in release workflows, we can adopt a similar solution to test it locally on mock instances.

Reply:
Thanks @LecrisUT and @mahendrapaipuri for this. I am afraid this is too unfamiliar to me. I am happy with the current process for releasing, i.e. editing the `CHANGELOG.md` and version files manually, and then adding a tag in the GitHub interface, so if you don't mind might left this change for later and revert to the previous state of `publish.yml` for now.